---
title: "`r gsub('.Rmd', ' ', gsub('_', ' ', knitr::current_input()))`"
author: "`r Sys.getenv('USER')`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: true
---

```{r setup}

library(scales) # for comma_format
library(tidyverse)
library(dplyr)
library(tidyr)
library(janitor)
library(cowplot)


```

## Summarize results of bulk processing for outlier lead accuracy project

This data comes from the CKCC1 outlier results.

## Identify input data

```{r}

# data_release_id <- "CKCC_outlier_lead_accuracy.results.2020_07_17"
# outlier_probability_output_CKCC_2020_12_03.tsv.gz

```

# Load data
```{r}

#gathered_results <- read_tsv(paste0("bulk_results.", data_release_id, ".tsv.gz")) %>%
#  rename(Mean_Accuracy = `Mean Accuracy`)

gathered_results_raw <- read_tsv("2020_12_03/outlier_probability_output_CKCC_2020_12_04.tsv.gz") 

gathered_results <- gathered_results_raw %>%
  select(Sample, Gene, Type, Expression, 
         Threshold, MEND_depth = `Mend depth`,
         effective_length = `Gene length`,
         Mean_Accuracy = `Probability of being an outlier`) %>%
  mutate(`Percent difference` = 100 * ((abs(Expression - Threshold)) / Expression))


```


# Add bin info
```{r}

expression_bin_boundaries <- c(0, 1, 3, 5, 7, 10, 20)
mend_depth_bin_boundaries <- c(0, seq(2, 46, 4), Inf)*1E6
length_bin_boundaries <- c(0, 100, 450, 800, 1500, 4000, Inf)


make_bin_names <- function(boundaries = 1:10){
  raw_names <- paste0(boundaries, "-", lead(boundaries)) %>% str_replace("-Inf", "+")
  raw_names[1:(length(boundaries)-1)]
}
  

expression_bin_names <- make_bin_names(expression_bin_boundaries)
mend_depth_bin_names <- make_bin_names(mend_depth_bin_boundaries/1e6)
length_bin_names <- make_bin_names(length_bin_boundaries)

bulk_results <- gathered_results %>%
  mutate(Expression_bin = cut(Expression, expression_bin_boundaries, 
                              labels = expression_bin_names,
                        include.lowest = TRUE),
         MEND_bin = cut(MEND_depth, mend_depth_bin_boundaries, labels = mend_depth_bin_names),
         length_bin = cut(effective_length, length_bin_boundaries, labels = length_bin_names,
                                      include.lowest = TRUE)
         )

```

# How many bins are possible? How many contain entries?
```{r}
print("How many bins are possible?")
(length(expression_bin_boundaries)-1) * (length(mend_depth_bin_boundaries)-1) * (length(length_bin_boundaries)-1)

bin_check <- bulk_results %>%
  group_by(Expression_bin, MEND_bin, length_bin) %>%
  summarize(n_entries_per_bin = n())

print("How many bins contain entries in this data?")
nrow(bin_check)

print("What is the range of the number of entries per bin?")
bin_check %>%
  pull(n_entries_per_bin) %>%
  summary

# What size are bins?
ggplot(bin_check) + 
  geom_histogram(aes(x=n_entries_per_bin)) +
  ggtitle("Entries per bin")
# 
# # Which bins have only one entry?
# subset(bin_check, n_entries_per_bin == 1)

print("How many bins have between 1 and 99 results?")
length(which(bin_check$n_entries_per_bin > 0 & bin_check$n_entries_per_bin < 100))

print("How many detected outlier genes total are being analyzed?")
sum(bin_check$n_entries_per_bin)

```

# Confirm the absence of NA values
```{r}

dim(bulk_results)
dim(is.na(bulk_results))

sum(is.na(bulk_results$Mean_Accuracy))

```

## Profile of MEND depths
```{r}
just_MEND_depths <- bulk_results %>% 
  select(Sample, MEND_depth)

ggplot(just_MEND_depths) +
  geom_histogram(aes(x=MEND_depth/1E6))

range(just_MEND_depths$MEND_depth/1e6)  

range(summary(just_MEND_depths$MEND_depth/1e6))  

```


## Number of outlier leads per sample

```{r}

# Histogram of number of outlier leads per sample (split by outlier type & fill is depth)

num_leads <- bulk_results %>% group_by(Sample, Type, MEND_bin) %>% tally()

plot1 <- ggplot(num_leads, aes(x=n, fill=as.factor(MEND_bin))) +
  geom_histogram() +
  facet_wrap(~Type) +
  ggtitle("Number of outlier leads per sample")

plot1

```

## Look at bin sizes

```{r}

# Make sure (depth, expression, length) bins are relatively similar sizes

all_bins <- bulk_results %>% tabyl(MEND_bin, Expression_bin, length_bin)

all_bins

```

## How many outliers had >95% accuracy?

```{r}

print("How many outliers had >95% accuracy?")
length(which(bulk_results$Mean_Accuracy > 0.95))

print("What percent of outliers had >95% accuracy?")
100*length(which(bulk_results$Mean_Accuracy > 0.95))/nrow(bulk_results)

```

## Spectrum of accuracies by depth, expression, and length

```{r}

plot2 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~MEND_bin) +
  ggtitle("Spectrum of accuracies by MEND depth")

plot2

plot3 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~Expression_bin) +
  ggtitle("Spectrum of accuracies by expression level")

plot3

plot4 <- ggplot(bulk_results, aes(x=Mean_Accuracy)) +
  geom_histogram() +
  facet_wrap(~length_bin) +
  ggtitle("Spectrum of accuracies by effective gene length")

plot4

```

# Spectrum of accuracies by outlier type (histograms for figure 1)

```{r}

bulk_results$Type <- factor(bulk_results$Type, levels = c("Up", "Down"))

acc_overview <- ggplot(bulk_results, aes(x=Mean_Accuracy, color=Type)) +
  geom_histogram(fill="white") +
  facet_wrap(~Type, scales="free") +
  scale_color_brewer(palette = "Set1") +
  theme_bw() +
  xlab("Accuracy") +
  ylab("Gene expression measurements") +
  ggtitle("Accuracies of up and down outlier gene expression measurements")

acc_overview
ggsave("../figures/accuracies_all.png", acc_overview)

```

# Distributions of accuracies for different expressions, depths, and lengths (boxplots for figure 1)

```{r}

figure1_data <- bulk_results %>%
  mutate(exp_gt5 = Expression >= 5,
         depth_gt20 = MEND_bin >= 20,
         length_gt1000 = effective_length >= 1000)

# Expression above or below 5

ggplot(subset(figure1_data, Type == "Up"), aes(x=exp_gt5, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Exp >= 5", "FALSE"="Exp < 5")) +
  ylab("Accuracy") +
  xlab("Expression above or below 5 log2(TPM+1)") +
  ggtitle("Accuracies of up outliers with high and low gene expression")

ggsave("..//figures/accuracies_up_exp.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=exp_gt5, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Exp >= 5", "FALSE"="Exp < 5")) +
  ylab("Accuracy") +
  xlab("Gene expression above or below 5 log2(TPM+1)") +
  ggtitle("Accuracies of down outliers with high and low gene expression")

ggsave("..//figures/accuracies_down_exp.png")

# Depth above or below 20

ggplot(subset(figure1_data, Type == "Up"), aes(x=depth_gt20, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Depth >= 20", "FALSE"="Depth < 20")) +
  ylab("Accuracy") +
  xlab("MEND depth above or below 20 million MEND reads") +
  ggtitle("Accuracies of up outliers with high and low MEND read depth")

ggsave("..//figures/accuracies_up_depth.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=depth_gt20, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Depth >= 20", "FALSE"="Depth < 20")) +
  ylab("Accuracy") +
  xlab("MEND depth above or below 20 million MEND reads") +
  ggtitle("Accuracies of down outliers with high and low MEND read depth")

ggsave("..//figures/accuracies_down_depth.png")

# Length above or below 1000

ggplot(subset(figure1_data, Type == "Up"), aes(x=length_gt1000, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Length >= 1000", "FALSE"="Length < 1000")) +
  ylab("Accuracy") +
  xlab("Gene length above or below 1000 base pairs") +
  ggtitle("Accuracies of up outliers with high and low effective gene length")

ggsave("..//figures/accuracies_up_length.png")

ggplot(subset(figure1_data, Type == "Down"), aes(x=length_gt1000, y=Mean_Accuracy)) +
  geom_boxplot() +
  coord_flip() +
  theme_bw() +
  scale_x_discrete(labels=c("TRUE"="Length >= 1000", "FALSE"="Length < 1000")) +
  ylab("Accuracy") +
  xlab("Gene length above or below 1000 base pairs") +
  ggtitle("Accuracies of down outliers with high and low effective gene length")

ggsave("..//figures/accuracies_down_length.png")

```


# StatBin2 function def
```{r}

# StatBin2 allows depiction of empty bins as blank instead of a horizontal line:
# https://stackoverflow.com/questions/57128090/remove-baseline-color-for-geom-histogram
StatBin2 <- ggproto(
  "StatBin2", 
  StatBin,
  compute_group = function (data, scales, binwidth = NULL, bins = NULL, 
                            center = NULL, boundary = NULL, 
                            closed = c("right", "left"), pad = FALSE, 
                            breaks = NULL, origin = NULL, right = NULL, 
                            drop = NULL, width = NULL) {
    if (!is.null(breaks)) {
      if (!scales$x$is_discrete()) {
        breaks <- scales$x$transform(breaks)
      }
      bins <- ggplot2:::bin_breaks(breaks, closed)
    }
    else if (!is.null(binwidth)) {
      if (is.function(binwidth)) {
        binwidth <- binwidth(data$x)
      }
      bins <- ggplot2:::bin_breaks_width(scales$x$dimension(), binwidth, 
                                         center = center, boundary = boundary, 
                                         closed = closed)
    }
    else {
      bins <- ggplot2:::bin_breaks_bins(scales$x$dimension(), bins, 
                                        center = center, boundary = boundary, 
                                        closed = closed)
    }
    res <- ggplot2:::bin_vector(data$x, bins, weight = data$weight, pad = pad)

    # drop 0-count bins completely before returning the dataframe
    res <- res[res$count > 0, ] 

    res
  })


```



# Look at the distribution of percent difference values to pick a split
```{r}

summary(bulk_results$`Percent difference`)
summary(bulk_results$MEND_depth)
summary(bulk_results$effective_length)
summary(bulk_results$Expression)

median(bulk_results$`Percent difference`)
median(bulk_results$MEND_depth)
median(bulk_results$effective_length)
median(bulk_results$Expression)

ggplot(bulk_results) +
  geom_point(aes(x=MEND_depth/1e6, y=Mean_Accuracy)) + 
  facet_wrap(~Type)


ggplot(bulk_results) +
  geom_point(aes(x=effective_length, y=Mean_Accuracy)) + 
  facet_wrap(~Type) +
  xlim(0, 25000)


ggplot(bulk_results) +
  geom_point(aes(x=Expression, y=Mean_Accuracy)) + 
  facet_wrap(~Type)



```


# Figure 1
```{r figure-1}

# to get hex codes from palette
# library(RColorBrewer)
# brewer.pal(12, "Set1")

type_colors = c(Up = "#E41A1C", Down = "#377EB8")

bulk_results$Type <- factor(bulk_results$Type, levels = c("Up", "Down"))

acc_overview <- ggplot(bulk_results, aes(x=Mean_Accuracy, color=Type)) +
  geom_histogram(fill="white", stat = StatBin2) +
  facet_wrap(~Type, scales="free") +
  theme_bw() +
  scale_color_manual(values = type_colors) +
  theme(legend.position="none") +
  xlab("") +
  ylab("Gene\nexpression\nmeasurements")

figure1_data_v2 <- bind_rows(
bulk_results %>%
  filter(Expression > quantile(Expression, 0.75)) %>%
  mutate(set = paste0("Expression>", round(quantile(Expression, 0.75), 1), " Log2(TPM+1)")),
bulk_results %>%
  filter(Expression < quantile(Expression, 0.25)) %>%
  mutate(set = paste0("Expression<", round(quantile(Expression, 0.25), 1), " Log2(TPM+1)")),
bulk_results %>%
  filter(MEND_depth > quantile(MEND_depth, 0.75)) %>%
  mutate(set = paste0("Depth>", round(quantile(MEND_depth, 0.75)/1e6), "M MEND reads")),
bulk_results %>%
  filter(MEND_depth < quantile(MEND_depth, 0.25)) %>%
  mutate(set = paste0("Depth<", round(quantile(MEND_depth, 0.25)/1e6), "M MEND reads")),
bulk_results %>%
  filter(effective_length > quantile(effective_length, 0.75)) %>%
  mutate(set = paste0("Length>", round(quantile(effective_length, 0.75)/1000, 1), "kb")),
bulk_results %>%
  filter(effective_length < quantile(effective_length, 0.25)) %>%
  mutate(set = paste0("Length<", round(quantile(effective_length, 0.25)/1000, 1), "kb")),
bulk_results %>%
  filter(`Percent difference` > quantile(`Percent difference`, 0.75)) %>%
  mutate(set = paste0("Difference>", round(quantile(`Percent difference`, 0.75), 1), "%")),
bulk_results %>%
  filter(`Percent difference` < quantile(`Percent difference`, 0.25)) %>%
  mutate(set = paste0("Difference<", round(quantile(`Percent difference`, 0.25), 1), "%"))
)

boxplot_n_labels <- figure1_data_v2 %>%
   group_by(set, Type) %>%
   summarize(set_type_label = paste0("n=", scales::label_comma()(n())))

all_boxplots <- ggplot(figure1_data_v2, aes(y=set, x=Mean_Accuracy, fill = Type)) +
  geom_boxplot(outlier.size = 0.5) +
  geom_text(data = boxplot_n_labels, 
            #aes(x=0.50, y=set, label = set_type_label), vjust = -1, size = 3) +
            aes(x=0, y=set, label = set_type_label), vjust = -0.5, hjust = 0, size = 3) +
  theme_bw() +
  scale_fill_manual(values = type_colors) +
  scale_color_manual(values = type_colors) +
  facet_wrap(~Type, ncol = 2) +
  theme(legend.position="none") +
  ylab("")

all_boxplots

combined_image <- plot_grid(acc_overview, all_boxplots, ncol = 1, rel_heights = c(1,2), align = "v",
                  labels = "AUTO")
combined_image

ggsave("../figures/fig1.png", combined_image)

```


Notes: the reason that, for up-outliers, difference == 100% has a median lower than difference < 2% is because  the difference is only 100 when the threshold is 0. Usually these are low expressed genes, so they get the max mean accuracy of 88% (or whatever the exact numbers are)


Random calculations for the paper

```{r}

# Average number of up and down outliers per sample

mean(subset(num_leads, Type == "Up")$n)
mean(subset(num_leads, Type == "Down")$n)

median(subset(num_leads, Type == "Up")$n)
median(subset(num_leads, Type == "Down")$n)

# Average accuracies of up and down outliers per sample

mean(subset(figure1_data, Type == "Up" & !is.na(Mean_Accuracy))$Mean_Accuracy)
mean(subset(figure1_data, Type == "Down" & !is.na(Mean_Accuracy))$Mean_Accuracy)

median(subset(figure1_data, Type == "Up" & !is.na(Mean_Accuracy))$Mean_Accuracy)
median(subset(figure1_data, Type == "Down" & !is.na(Mean_Accuracy))$Mean_Accuracy)

# Fraction of accuracies above 95%

count(subset(figure1_data, Type == "Up" & Mean_Accuracy>95)) /
  count(subset(figure1_data, Type == "Up"))

count(subset(figure1_data, Type == "Down" & Mean_Accuracy>95)) /
  count(subset(figure1_data, Type == "Down"))

count(subset(figure1_data, Mean_Accuracy>95)) /
  count(figure1_data)

# the bump in distribution due to the 0-1 crowd

median(subset(figure1_data, Expression_bin == "0-1")$Mean_Accuracy)

```



# SessionInfo
```{r}
library(pander) # to make session info
pander(sessionInfo(), compact = FALSE)
```

