---
title: "`r gsub('.Rmd', ' ', gsub('_', ' ', knitr::current_input()))`"
author: "`r Sys.getenv('USER')`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
---

```{r setup}

library(tidyverse)
library(dplyr)
library(tidyr)
library(janitor)
library(cowplot)
library(scales) # for comma_format

```

## Summarize results of bulk processing for outlier lead accuracy project

This data comes from the CKCC1 outlier results.

## Identify input data

```{r}

data_release_id <- "CKCC_outlier_lead_accuracy.results.2020_07_17"

```

# Load data
```{r}

gathered_results <- read_tsv(paste0("bulk_results.", data_release_id, ".tsv.gz")) %>%
  rename(Mean_Accuracy = `Mean Accuracy`)

```


# Add bin info
```{r}

expression_bin_boundaries <- c(0, 1, 3, 5, 7, 10, 20)
mend_depth_bin_boundaries <- c(seq(0, 44, 4), Inf)*1E6
length_bin_boundaries <- c(seq(0, 3000, 300), Inf)

make_bin_names <- function(boundaries = 1:10){
  raw_names <- paste0(boundaries, "-", lead(boundaries)) %>% str_replace("-Inf", "+")
  raw_names[1:(length(boundaries)-1)]
}
  
expression_bin_names <- make_bin_names(expression_bin_boundaries)
mend_depth_bin_names <- make_bin_names(mend_depth_bin_boundaries/1e6)
length_bin_names <- make_bin_names(length_bin_boundaries)

bulk_results <- gathered_results %>%
  mutate(Expression_bin = cut(Expression, expression_bin_boundaries, 
                              labels = expression_bin_names,
                        include.lowest = TRUE),
         MEND_bin = cut(MEND_depth, mend_depth_bin_boundaries, labels = mend_depth_bin_names),
         length_bin = cut(effective_length, length_bin_boundaries, labels = length_bin_names,
                                      include.lowest = TRUE),
         Type = factor(Type, levels = c("Up", "Down"))) %>%
  group_by(Type) %>%
  mutate(Type_label = paste0(as.character(Type), " (n=", as.character(label_comma()(n())), ")")) %>%
  ungroup()
  

bulk_results$Type_label = factor(bulk_results$Type_label, levels = sort(unique(as.character(bulk_results$Type_label)), decreasing = TRUE))

```


# Spectrum of accuracies by outlier type (histograms for figure 1)

```{r}

acc_overview <- ggplot(bulk_results, aes(x=Mean_Accuracy, color=Type)) +
  geom_histogram(fill="white") +
  facet_wrap(~Type_label, scales="free") +
  scale_color_brewer(palette = "Set1") +
  theme_bw() +
  xlab("Accuracy") +
  ylab("Gene\nexpression\nmeasurements") +
  ggtitle("Accuracies of up and down outlier gene expression measurements") + 
  theme(legend.position="none")

acc_overview
ggsave("../figures/accuracies_all.png", acc_overview)

```

# Distributions of accuracies for different expressions, depths, and lengths (boxplots for figure 1)

## StatBin2 function def
```{r}

# StatBin2 allows depiction of empty bins as blank instead of a horizontal line:
# https://stackoverflow.com/questions/57128090/remove-baseline-color-for-geom-histogram
StatBin2 <- ggproto(
  "StatBin2", 
  StatBin,
  compute_group = function (data, scales, binwidth = NULL, bins = NULL, 
                            center = NULL, boundary = NULL, 
                            closed = c("right", "left"), pad = FALSE, 
                            breaks = NULL, origin = NULL, right = NULL, 
                            drop = NULL, width = NULL) {
    if (!is.null(breaks)) {
      if (!scales$x$is_discrete()) {
        breaks <- scales$x$transform(breaks)
      }
      bins <- ggplot2:::bin_breaks(breaks, closed)
    }
    else if (!is.null(binwidth)) {
      if (is.function(binwidth)) {
        binwidth <- binwidth(data$x)
      }
      bins <- ggplot2:::bin_breaks_width(scales$x$dimension(), binwidth, 
                                         center = center, boundary = boundary, 
                                         closed = closed)
    }
    else {
      bins <- ggplot2:::bin_breaks_bins(scales$x$dimension(), bins, 
                                        center = center, boundary = boundary, 
                                        closed = closed)
    }
    res <- ggplot2:::bin_vector(data$x, bins, weight = data$weight, pad = pad)

    # drop 0-count bins completely before returning the dataframe
    res <- res[res$count > 0, ] 

    res
  })


```



## Prepare data for Figure 1
```{r figure-1}
# to get hex codes from palette
# library(RColorBrewer)
# brewer.pal(12, "Set1")

type_colors = c(Up = "#E41A1C", Down = "#377EB8")

bulk_results$Type <- factor(bulk_results$Type, levels = c("Up", "Down"))

split_by_quartile <- function(bulk_results, col_name_fn = Expression, this_suffix = " Log2(TPM+1)"){
  col_name_text <- rlang::ensym(col_name_fn)
  # Upper quartile
 bulk_results_bit <- lapply(c("Up", "Down"), function(this_type){
   these_bulk_results <- bulk_results %>% filter(Type == this_type)
    this_threshold <- these_bulk_results %>% summarize(this_q = quantile(!!col_name_text, 0.75)) %>% pull(this_q) %>% unique
    bulk_results_upper_bit <- these_bulk_results %>% 
      filter(!!col_name_text > this_threshold) %>%
      mutate(set = paste0(rlang::as_label(col_name_text),  "_UQ"),
             set_type_label = paste0(">", round(this_threshold, 1), this_suffix, ", n=", label_comma()(n())))
  # Lower quartile
    this_threshold <- these_bulk_results %>% summarize(this_q = quantile(!!col_name_text, 0.25)) %>% pull(this_q) %>% unique
    bulk_results_lower_bit <- these_bulk_results %>% 
      filter(!!col_name_text < this_threshold) %>%
      mutate(set = paste0(rlang::as_label(col_name_text), "_LQ"),
             set_type_label = paste0("<", round(this_threshold, 1), this_suffix, ", n=", label_comma()(n())))
   these_bits <- bind_rows(bulk_results_upper_bit, bulk_results_lower_bit) 
  }) %>% bind_rows()
  return(bulk_results_bit)
}

bulk_results_for_fig1 <- 
  bulk_results %>% 
  mutate(Depth = MEND_depth/1e6,
         Length = effective_length/1e3,
         Difference = `Percent difference`)

figure1_data_v2 <- bind_rows(
split_by_quartile(bulk_results_for_fig1, col_name = Expression, this_suffix = " Log2(TPM+1)"),
split_by_quartile(bulk_results_for_fig1, col_name = Depth, this_suffix = "M MEND reads"),
split_by_quartile(bulk_results_for_fig1, col_name = Length, this_suffix = "kb"),
split_by_quartile(bulk_results_for_fig1, col_name = Difference, this_suffix = "%"),
)

sum(bulk_results_for_fig1$Length>2.4)
sum(bulk_results_for_fig1$Type=="Up")
tabyl(bulk_results_for_fig1$Type)
```

## Generate figures
```{r figure-1g}
boxplot_n_labels <- figure1_data_v2 %>%
   select(set, set_type_label, Type, Type_label) %>%
  distinct()


all_boxplots <- ggplot(figure1_data_v2, aes(y=set, x=Mean_Accuracy, fill = Type)) +
  geom_boxplot(outlier.size = 0.5) +
  geom_text(data = boxplot_n_labels, 
            aes(x=5, y=set, label = set_type_label), hjust = 0, vjust = -1, size = 3) +
  theme_bw() +
  scale_fill_manual(values = type_colors) +
  scale_color_manual(values = type_colors) +
  facet_wrap(~Type, ncol = 2) +
  theme(legend.position="none") +
  ylab("")

all_boxplots
```


```{r figure-1f, fig.height=7, fig.width=7}
combined_image <- plot_grid(acc_overview, all_boxplots, ncol = 1, rel_heights = c(1,2), align = "v",
                  labels = "AUTO")
combined_image

ggsave("../figures/fig1.png", combined_image)

```


Notes: the reason that, for up-outliers, the upper quartile of differences has a median lower than the lower quartile of differences because of the many genes with low expression have a max mean accuracy of 88% (or whatever the exact numbers are)

```{r}


summary(subset(bulk_results, Type == "Up")$`Percent difference`)
        
        summary(bulk_results$`Percent difference`[!bulk_results$`Percent difference` == Inf])

        
        quantile(bulk_results$`Percent difference`, 0.75)
  
        round(quantile(bulk_results$`Percent difference`, 0.75), 1)
        
                      
        bulk_results %>%
  filter(`Percent difference` > quantile(`Percent difference`, 0.75),
         Type == "Up") %>%
  mutate(set = paste0("Difference>", round(quantile(`Percent difference`, 0.75), 1), "%"))
        
```

Random calculations for the paper

```{r}

# changed figure1_data to  bulk_results in this section

num_leads <- bulk_results %>% group_by(Sample, Type, MEND_bin) %>% tally()

# Average number of up and down outliers per sample

mean(subset(num_leads, Type == "Up")$n)
mean(subset(num_leads, Type == "Down")$n)

median(subset(num_leads, Type == "Up")$n)
median(subset(num_leads, Type == "Down")$n)

# Average accuracies of up and down outliers per sample

mean(subset(bulk_results, Type == "Up" & !is.na(Mean_Accuracy))$Mean_Accuracy)
mean(subset(bulk_results, Type == "Down" & !is.na(Mean_Accuracy))$Mean_Accuracy)

median(subset(bulk_results, Type == "Up" & !is.na(Mean_Accuracy))$Mean_Accuracy)
median(subset(bulk_results, Type == "Down" & !is.na(Mean_Accuracy))$Mean_Accuracy)

# Fraction of accuracies above 95%

count(subset(bulk_results, Type == "Up" & Mean_Accuracy>95)) /
  count(subset(bulk_results, Type == "Up"))

count(subset(bulk_results, Type == "Down" & Mean_Accuracy>95)) /
  count(subset(bulk_results, Type == "Down"))

count(subset(bulk_results, Mean_Accuracy>95)) /
  count(bulk_results)

# the bump in distribution due to the 0-1 crowd

median(subset(bulk_results, Expression_bin == "0-1")$Mean_Accuracy)

```

